### PR TITLE
Strings concept and Log Line Parser exercise

### LINKED ISSUE #94: String and string interpolation concept
next concept: strings.

### PR DESCRIPTION
**Pull request by glennj**
Closes #94

### CODE REVIEW

File: concepts/strings/about.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+Within a string, use the backslash character to embed "special" characters:
+
+- `\"` a literal quotation mark,
+- `\\` a literal backslash,
+- whitespace characters `\n`, `\t`, `\r`, `\f`, `\b`
+- unicode characters `\uAAAA` where `A` is a hex digit
+
+## String length
+
+To find the number of characters: [`length`][length]
```
Comment (IsaacG) [inline concepts/strings/about.md:19]:
```suggestion
To find the number of characters, use the [`length`][length] function
```

File: concepts/strings/about.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+```sh
+$ jq -cn --args  '$ARGS.positional[] | length' "Hello world!" "❄🌡🤧🤒🏥🕰😀"
+12
+7
+```
+
+To find the number of actual _bytes_: [`utf8bytelength`][utf8bytelength]
```
Comment (IsaacG) [inline concepts/strings/about.md:26]:
```suggestion
To find the number of actual _bytes_, use [`utf8bytelength`][utf8bytelength]
```

File: concepts/strings/about.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+```jq
+"abcdefghij"[3:6]   # => "def"
+"abcdefghij"[3:]    # => "defghij"
+"abcdefghij"[:-2]   # => "abcdefgh"
+```
+
+The first number is the (zero-based) start index (or 0 if absent).
+The second number is the end index _plus one_ (or end of string if absent).
+That is a bit confusing, but it corresponds with the `range/2` function where the 2 parameters are the range start (inclusive) and range end (exclusive).
+## Concatenation
```
Comment (IsaacG) [inline concepts/strings/about.md:46]:
```suggestion

## Concatenation
```

File: concepts/strings/about.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    ```jq
+    "Hi friend 😀" | explode   # => [72,105,32,102,114,105,101,110,100,32,128512]
+    ```
+
+    [`implode`][implode] is the inverse of [`explode`][explode].
+
+## Find the index of a substring
+
+Use the [`index/1`][index/1] function: the index is zero-based.
```
Comment (IsaacG) [inline concepts/strings/about.md:94]:
```suggestion
Use the [`index/1`][index/1] function; the index is zero-based.
```

File: concepts/strings/about.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+```jq
+"hello" | index("el")'   # => 1
+```
+
+If the substring is not in the string, the result is `null`
+```jq
+"hello" | index("elk")   # => null
+```
+
+Also useful: [`rindex/1`][index/1] and [`indices`][indices]
```
Comment (IsaacG) [inline concepts/strings/about.md:105]:
```suggestion
Also useful are the functions [`rindex/1`][index/1] and [`indices`][indices].
```

File: concepts/strings/about.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+Within a string, the sequence `\(expression)` will embed the _result_ of that expression [into the string][interpolate]:
+
+```jq
+"The current datetime is \(now | strflocaltime("%c"))"   # => "The current datetime is Wed Nov 16 17:06:33 2022"
+```
+
+## Case conversion
+
+Use the [`ascii_downcase`][ascii_downcase] and [`ascii_upcase`][ascii_downcase] functions.
```
Comment (IsaacG) [inline concepts/strings/about.md:117]:
```suggestion
Use the [`ascii_downcase`][ascii_downcase] and [`ascii_upcase`][ascii_downcase] functions to change the case of a string.
```

File: concepts/strings/about.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+## Cast a string to a number
+
+Use the [`tonumber`][tonumber] function.
```
Comment (IsaacG) [inline concepts/strings/about.md:121]:
```suggestion
Use the [`tonumber`][tonumber] function to convert a string to a number.
```

File: exercises/concept/log-line-parser/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# Instructions
+
+In this exercise you'll be processing log-lines.
```
Comment (IsaacG) [inline exercises/concept/log-line-parser/.docs/instructions.md:3]:
```suggestion
In this exercise you'll be processing log entries.
```

File: exercises/concept/log-line-parser/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+Each log line is a string formatted as follows: `"[<LEVEL>]: <MESSAGE>"`.
+
+There are three different log levels:
+
+- `INFO`
+- `WARNING`
+- `ERROR`
+
+You have three tasks, each of which will take a log line and ask you to do something with it.
+There are lots of ways to solve these tasks - choose your favourite methods from the examples above and see what you can come up with.
```
Comment (IsaacG) [inline exercises/concept/log-line-parser/.docs/instructions.md:14]:
```suggestion
There are lots of ways to solve these tasks; choose your favourite methods from the examples above and see what you can come up with.
```

File: exercises/concept/log-line-parser/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+## 1. Get message from a log line
+
+Implement the `LogLineParser#message` method to return a log line's message:
+
+```ruby
```
Comment (IsaacG) [inline exercises/concept/log-line-parser/.docs/instructions.md:20]: Do you want to mention you're showing Ruby code prior to showing Ruby-formatted code?

Comment (glennj) [reply exercises/concept/log-line-parser/.docs/instructions.md:20]: I copied the exercise over from the ruby track: not quite finished with it yet. Thanks for the sharp eyed review!

File: exercises/concept/lasagna/.meta/config.json [hunk 11]
```diff
@@ -1,9 +1,17 @@
 {
```
Comment (glennj) [inline exercises/concept/lasagna/.meta/config.json:1]: This file changed due to a "global" `configlet fmt` I inadvertantly performed.

File: exercises/concept/shopping/.meta/config.json [hunk 12]
```diff
@@ -1,10 +1,20 @@
 {
```
Comment (glennj) [inline exercises/concept/shopping/.meta/config.json:1]: This file changed due to a "global" `configlet fmt` I inadvertantly performed.

**Comment by glennj** [comment]
Man this is a mess. I'll recreate.

**Pull request closed without merging**